File: Services/ServerPingService.cs
using System.Buffers.Binary;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace GlacierLauncher.Services;

/// <summary>Result of a server status query. <see cref="Online"/> is false on any failure/timeout.</summary>
public sealed record ServerStatus(
    bool Online, int PlayersOnline, int PlayersMax, string Motd, string Version, long LatencyMs)
{
    public static readonly ServerStatus Offline = new(false, 0, 0, "", "", 0);
}

/// <summary>
/// Pings Minecraft servers for live status (player counts, MOTD, latency) without
/// any third-party API: Bedrock via the RakNet unconnected-ping (UDP), Java via the
/// Server List Ping handshake (TCP). Every failure path resolves to <see
/// cref="ServerStatus.Offline"/> so callers never have to catch.
/// </summary>
public static class ServerPingService
{
    private static readonly byte[] RakNetMagic =
    {
        0x00, 0xff, 0xff, 0x00, 0xfe, 0xfe, 0xfe, 0xfe,
        0xfd, 0xfd, 0xfd, 0xfd, 0x12, 0x34, 0x56, 0x78
    };

    /// <summary>Picks Java SLP for port 25565, otherwise Bedrock RakNet.</summary>
    public static Task<ServerStatus> PingAsync(string host, int port, int timeoutMs = 2500) =>
        port == 25565 ? PingJavaAsync(host, port, timeoutMs) : PingBedrockAsync(host, port, timeoutMs);

    // ── Bedrock (RakNet unconnected ping over UDP) ───────────────────────────
    public static async Task<ServerStatus> PingBedrockAsync(string host, int port, int timeoutMs = 2500)
    {
        try
        {
            using var udp = new UdpClient();
            var sw = Stopwatch.StartNew();

            var packet = new byte[1 + 8 + 16 + 8];
            packet[0] = 0x01;                                                   // ID_UNCONNECTED_PING
            BinaryPrimitives.WriteInt64BigEndian(packet.AsSpan(1, 8), sw.ElapsedMilliseconds);
            RakNetMagic.CopyTo(packet.AsSpan(9, 16));
            BinaryPrimitives.WriteInt64BigEndian(packet.AsSpan(25, 8), 0x0102030405060708);

            await udp.SendAsync(packet, packet.Length, host, port).ConfigureAwait(false);

            var recv = udp.ReceiveAsync();
            if (await Task.WhenAny(recv, Task.Delay(timeoutMs)).ConfigureAwait(false) != recv)
                return ServerStatus.Offline;
            sw.Stop();

            var data = recv.Result.Buffer;                                      // 0x1c | time(8) | guid(8) | magic(16) | len(2) | id-string
            if (data.Length < 35 || data[0] != 0x1c) return ServerStatus.Offline;

            int strLen = BinaryPrimitives.ReadUInt16BigEndian(data.AsSpan(33, 2));
            var id = Encoding.UTF8.GetString(data, 35, Math.Min(strLen, data.Length - 35));

            // edition;motd;protocol;version;online;max;guid;motd2;gamemode;...
            var f = id.Split(';');
            return new ServerStatus(
                Online:        true,
                PlayersOnline: f.Length > 4 && int.TryParse(f[4], out var o) ? o : 0,
                PlayersMax:    f.Length > 5 && int.TryParse(f[5], out var m) ? m : 0,
                Motd:          f.Length > 1 ? f[1] : "",
                Version:       f.Length > 3 ? f[3] : "",
                LatencyMs:     sw.ElapsedMilliseconds);
        }
        catch { return ServerStatus.Offline; }
    }

    // ── Java (Server List Ping over TCP) ─────────────────────────────────────
    public static async Task<ServerStatus> PingJavaAsync(string host, int port, int timeoutMs = 2500)
    {
        using var cts = new CancellationTokenSource(timeoutMs);
        var ct = cts.Token;
        try
        {
            using var tcp = new TcpClient();
            var sw = Stopwatch.StartNew();
            await tcp.ConnectAsync(host, port, ct).ConfigureAwait(false);
            await using var stream = tcp.GetStream();

            using var hs = new MemoryStream();
            WriteVarInt(hs, 0x00);                       // handshake packet id
            WriteVarInt(hs, -1);                         // protocol version (-1 = "any")
            WriteString(hs, host);
            hs.WriteByte((byte)(port >> 8));
            hs.WriteByte((byte)(port & 0xFF));           // port as unsigned short, big-endian
            WriteVarInt(hs, 1);                          // next state = status
            await WritePacketAsync(stream, hs.ToArray(), ct).ConfigureAwait(false);

            await WritePacketAsync(stream, new byte[] { 0x00 }, ct).ConfigureAwait(false); // status request

            _ = await ReadVarIntAsync(stream, ct).ConfigureAwait(false);        // total length
            _ = await ReadVarIntAsync(stream, ct).ConfigureAwait(false);        // packet id (0x00)
            int jsonLen = await ReadVarIntAsync(stream, ct).ConfigureAwait(false);
            var json = await ReadExactAsync(stream, jsonLen, ct).ConfigureAwait(false);
            sw.Stop();

            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            int online = 0, max = 0; string ver = "", motd = "";
            if (root.TryGetProperty("players", out var pl))
            {
                if (pl.TryGetProperty("online", out var o)) online = o.GetInt32();
                if (pl.TryGetProperty("max", out var m))    max    = m.GetInt32();
            }
            if (root.TryGetProperty("version", out var v) && v.TryGetProperty("name", out var vn))
                ver = vn.GetString() ?? "";
            if (root.TryGetProperty("description", out var d))
                motd = ExtractMotd(d);

            return new ServerStatus(true, online, max, motd, ver, sw.ElapsedMilliseconds);
        }
        catch { return ServerStatus.Offline; }
    }

    // ── VarInt / packet helpers ──────────────────────────────────────────────
    private static void WriteVarInt(Stream s, int value)
    {
        uint v = (uint)value;
        do
        {
            byte b = (byte)(v & 0x7F);
            v >>= 7;
            if (v != 0) b |= 0x80;
            s.WriteByte(b);
        } while (v != 0);
    }

    private static void WriteString(Stream s, string str)
    {
        var bytes = Encoding.UTF8.GetBytes(str);
        WriteVarInt(s, bytes.Length);
        s.Write(bytes, 0, bytes.Length);
    }

    private static async Task WritePacketAsync(Stream s, byte[] data, CancellationToken ct)
    {
        using var ms = new MemoryStream();
        WriteVarInt(ms, data.Length);
        ms.Write(data, 0, data.Length);
        var arr = ms.ToArray();
        await s.WriteAsync(arr, ct).ConfigureAwait(false);
    }

    private static async Task<int> ReadVarIntAsync(Stream s, CancellationToken ct)
    {
        int numRead = 0, result = 0; byte read;
        do
        {
            var one = new byte[1];
            if (await s.ReadAsync(one.AsMemory(0, 1), ct).ConfigureAwait(false) == 0)
                throw new EndOfStreamException();
            read = one[0];
            result |= (read & 0x7F) << (7 * numRead);
            if (++numRead > 5) throw new InvalidDataException("VarInt too big");
        } while ((read & 0x80) != 0);
        return result;
    }

    private static async Task<byte[]> ReadExactAsync(Stream s, int len, CancellationToken ct)
    {
        if (len < 0 || len > 1 << 20) throw new InvalidDataException("Bad length");
        var buf = new byte[len];
        int off = 0;
        while (off < len)
        {
            int n = await s.ReadAsync(buf.AsMemory(off, len - off), ct).ConfigureAwait(false);
            if (n == 0) throw new EndOfStreamException();
            off += n;
        }
        return buf;
    }

    private static string ExtractMotd(JsonElement d)
    {
        if (d.ValueKind == JsonValueKind.String) return d.GetString() ?? "";
        var sb = new StringBuilder();
        Walk(d, sb);
        return sb.ToString();

        static void Walk(JsonElement e, StringBuilder sb)
        {
            if (e.ValueKind == JsonValueKind.String) { sb.Append(e.GetString()); return; }
            if (e.ValueKind != JsonValueKind.Object) return;
            if (e.TryGetProperty("text", out var t)) sb.Append(t.GetString());
            if (e.TryGetProperty("extra", out var ex) && ex.ValueKind == JsonValueKind.Array)
                foreach (var c in ex.EnumerateArray()) Walk(c, sb);
        }
    }
}

File: android/app/src/main/java/xyz/glacierclient/launcher/service/ServerPingService.kt
package xyz.glacierclient.launcher.service

import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.InputStream
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

/**
 * Android analogue of Services/ServerPingService.cs. Same two wire protocols,
 * same "every failure resolves to offline" contract — this is called from a
 * JS `setInterval`-driven refresh over N saved/suggested servers, so a single
 * bad host must never throw and abort the batch, just report offline.
 *
 * Runs on whatever thread the @JavascriptInterface call lands on (WebView's
 * background JS-bridge thread, not the UI thread — same as every other
 * blocking-network bridge method in this app, e.g. BedrockVersionService's
 * downloadApk), so a plain blocking socket with a short timeout is fine; no
 * coroutine/executor plumbing needed.
 */
object ServerPingService {

    private val RAKNET_MAGIC = byteArrayOf(
        0x00, -0x01, -0x01, 0x00, -0x02, -0x02, -0x02, -0x02,
        -0x03, -0x03, -0x03, -0x03, 0x12, 0x34, 0x56, 0x78
    )

    private fun offline(): JSONObject = JSONObject()
        .put("online", false).put("playersOnline", 0).put("playersMax", 0)
        .put("motd", "").put("version", "").put("latencyMs", 0)

    /** Picks Java SLP for port 25565, otherwise Bedrock RakNet — same rule as the Windows service. */
    fun ping(host: String, port: Int, timeoutMs: Int = 2500): String =
        (if (port == 25565) pingJava(host, port, timeoutMs) else pingBedrock(host, port, timeoutMs)).toString()

    // ── Bedrock (RakNet unconnected ping over UDP) ────────────────────────
    private fun pingBedrock(host: String, port: Int, timeoutMs: Int): JSONObject {
        return try {
            DatagramSocket().use { socket ->
                socket.soTimeout = timeoutMs
                val start = System.nanoTime()

                val packet = ByteBuffer.allocate(1 + 8 + 16 + 8).order(ByteOrder.BIG_ENDIAN)
                packet.put(0x01)                          // ID_UNCONNECTED_PING
                packet.putLong(0L)                         // client send time (unused by us)
                packet.put(RAKNET_MAGIC)
                packet.putLong(0x0102030405060708L)        // client GUID
                val out = DatagramPacket(packet.array(), packet.array().size, InetSocketAddress(host, port))
                socket.send(out)

                val buf = ByteArray(1500)
                val inPkt = DatagramPacket(buf, buf.size)
                socket.receive(inPkt)                       // throws SocketTimeoutException on timeout
                val elapsedMs = (System.nanoTime() - start) / 1_000_000

                val data = inPkt.data
                // 0x1c | time(8) | guid(8) | magic(16) | idLen(2) | idString
                if (inPkt.length < 35 || data[0] != 0x1c.toByte()) return offline()
                val idLen = ((data[33].toInt() and 0xFF) shl 8) or (data[34].toInt() and 0xFF)
                val avail = (inPkt.length - 35).coerceAtLeast(0)
                val id = String(data, 35, minOf(idLen, avail), StandardCharsets.UTF_8)

                // edition;motd;protocol;version;online;max;guid;motd2;gamemode;...
                val f = id.split(";")
                JSONObject()
                    .put("online", true)
                    .put("playersOnline", f.getOrNull(4)?.toIntOrNull() ?: 0)
                    .put("playersMax", f.getOrNull(5)?.toIntOrNull() ?: 0)
                    .put("motd", f.getOrNull(1) ?: "")
                    .put("version", f.getOrNull(3) ?: "")
                    .put("latencyMs", elapsedMs)
            }
        } catch (_: Exception) {
            offline()
        }
    }

    // ── Java (Server List Ping over TCP) ──────────────────────────────────
    private fun pingJava(host: String, port: Int, timeoutMs: Int): JSONObject {
        return try {
            Socket().use { socket ->
                val start = System.nanoTime()
                socket.connect(InetSocketAddress(host, port), timeoutMs)
                socket.soTimeout = timeoutMs
                val out = DataOutputStream(socket.getOutputStream())
                val input = socket.getInputStream()

                val handshake = ByteArrayOutputStream().also { hs ->
                    writeVarInt(hs, 0x00)                  // handshake packet id
                    writeVarInt(hs, -1)                     // protocol version = "any"
                    writeString(hs, host)
                    hs.write((port shr 8) and 0xFF)
                    hs.write(port and 0xFF)                 // port, unsigned short big-endian
                    writeVarInt(hs, 1)                       // next state = status
                }
                writePacket(out, handshake.toByteArray())
                writePacket(out, byteArrayOf(0x00))          // status request

                readVarInt(input)                            // total packet length
                readVarInt(input)                            // packet id (0x00)
                val jsonLen = readVarInt(input)
                val jsonBytes = readExact(input, jsonLen)
                val elapsedMs = (System.nanoTime() - start) / 1_000_000

                val root = JSONObject(String(jsonBytes, StandardCharsets.UTF_8))
                val players = root.optJSONObject("players")
                val version = root.optJSONObject("version")
                val description = root.opt("description")
                val motd = when (description) {
                    is JSONObject -> description.optString("text", "")
                    is String -> description
                    else -> ""
                }

                JSONObject()
                    .put("online", true)
                    .put("playersOnline", players?.optInt("online", 0) ?: 0)
                    .put("playersMax", players?.optInt("max", 0) ?: 0)
                    .put("motd", motd)
                    .put("version", version?.optString("name", "") ?: "")
                    .put("latencyMs", elapsedMs)
            }
        } catch (_: Exception) {
            offline()
        }
    }

    // ── VarInt / packet helpers (mirrors the Windows service's own) ───────
    private fun writeVarInt(out: ByteArrayOutputStream, value: Int) {
        var v = value
        do {
            var b = v and 0x7F
            v = v ushr 7
            if (v != 0) b = b or 0x80
            out.write(b)
        } while (v != 0)
    }

    private fun writeString(out: ByteArrayOutputStream, s: String) {
        val bytes = s.toByteArray(StandardCharsets.UTF_8)
        writeVarInt(out, bytes.size)
        out.write(bytes)
    }

    private fun writePacket(out: DataOutputStream, payload: ByteArray) {
        val lenBuf = ByteArrayOutputStream()
        writeVarInt(lenBuf, payload.size)
        out.write(lenBuf.toByteArray())
        out.write(payload)
        out.flush()
    }

    private fun readVarInt(input: InputStream): Int {
        var result = 0
        var shift = 0
        while (true) {
            val b = input.read()
            if (b == -1) throw java.io.EOFException("stream closed mid-varint")
            result = result or ((b and 0x7F) shl shift)
            if (b and 0x80 == 0) break
            shift += 7
            if (shift >= 35) throw IllegalStateException("VarInt too long")
        }
        return result
    }

    private fun readExact(input: InputStream, length: Int): ByteArray {
        val buf = ByteArray(length)
        var off = 0
        while (off < length) {
            val n = input.read(buf, off, length - off)
            if (n == -1) throw java.io.EOFException("stream closed before $length bytes read")
            off += n
        }
        return buf
    }
}
